pkg/controller/seed-controller-manager/constraint-controller/controller.go

```diff
@@ -216,15 +230,24 @@ func (r *reconciler) patchFinalizer(ctx context.Context, constraint *kubermaticv
 
 func (r *reconciler) reconcile(ctx context.Context, constraint *kubermaticv1.Constraint, log *zap.SugaredLogger) error {
 
+	clusterList, unwantedClusterList, err := r.getClustersForConstraint(ctx, constraint)
+	if err != nil {
+		return fmt.Errorf("failed listing clusters: %w", err)
+	}
+
 	// constraint deletion
 	if !constraint.DeletionTimestamp.IsZero() {
+
 		if !kuberneteshelper.HasFinalizer(constraint, finalizer) {
 			return nil
 		}
 
-		if err := r.cleanupConstraint(ctx, log, constraint); err != nil {
-			return err
+		if clusterList != nil {
```

Comment:
you dont need to check if a list is nil, the iteration in `syncAllClustersNS` will work fine.

Same for the checks below :)

Reply:
right